File: src/main/java/me/dreamerzero/vlobby/commands/LobbyCommand.java
package me.dreamerzero.vlobby.commands;

import java.util.Arrays;

import com.mojang.brigadier.builder.LiteralArgumentBuilder;
import com.mojang.brigadier.tree.LiteralCommandNode;
import com.velocitypowered.api.command.BrigadierCommand;
import com.velocitypowered.api.command.CommandManager;
import com.velocitypowered.api.command.CommandMeta;
import com.velocitypowered.api.command.CommandSource;
import com.velocitypowered.api.permission.Tristate;
import com.velocitypowered.api.proxy.Player;
import com.velocitypowered.api.proxy.ProxyServer;
import com.velocitypowered.api.proxy.server.RegisteredServer;

import me.dreamerzero.vlobby.VLobby;
import me.dreamerzero.vlobby.utils.MessageUtils;
import me.dreamerzero.vlobby.utils.ServerUtils;
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder;

public final class LobbyCommand {
    private LobbyCommand(){}
    public static void loadCommands(VLobby plugin){
        ProxyServer proxy = plugin.getProxy();
        String[] commands = plugin.getConfig().getCommands();
        if(commands.length == 0){
            return;
        }
        LiteralCommandNode<CommandSource> lobbyCommand = LiteralArgumentBuilder.<CommandSource>literal(commands[0])
            .requires(src -> src.getPermissionValue("vlobby.command") != Tristate.FALSE)
            .executes(cmd -> {
                CommandSource source = cmd.getSource();
                if(!(source instanceof Player)){
                    plugin.getLogger().info(plugin.getConfig().getMessages().getConsoleMessage());
                    return 1;
                }
                Player player = (Player)source;
                RegisteredServer lobbyToSend = ServerUtils.getConfigServer(plugin);
                if(lobbyToSend == null){
                    plugin.getLogger().error("Cannot found lobby server to send");
                    MessageUtils.sendMessage(source, plugin.getConfig().getMessages().getNotAvailableServerMessage());
                    return 1;
                }
                player.createConnectionRequest(lobbyToSend).connect().thenAccept(result -> {
                    if(!result.isSuccessful()){
                        plugin.getLogger().error("Cannot connect to server: {}", lobbyToSend.getServerInfo().getName());
                        MessageUtils.sendMessage(
                            player,
                            plugin.getConfig().getMessages().getErrorMessage(),
                            Placeholder.unparsed("server", lobbyToSend.getServerInfo().getName())
                        );
                    } else {
                        MessageUtils.sendMessage(
                            player,
                            plugin.getConfig().getMessages().getSuccesfullyMessage(),
                            Placeholder.unparsed("server", lobbyToSend.getServerInfo().getName())
                        );
                    }
                });
                return 1;
            })
            .build();
        BrigadierCommand bCommand = new BrigadierCommand(lobbyCommand);
        CommandManager manager = proxy.getCommandManager();
        CommandMeta.Builder builder = manager.metaBuilder(bCommand).plugin(plugin);
        if(commands.length > 1){
            builder.aliases(Arrays.copyOfRange(commands, 1, commands.length-1));
        }
        manager.register(builder.build(), bCommand);
    }
}

File: src/main/kotlin/me/adrian3d/vlobby/commands/LobbyCommand.kt
package me.adrian3d.vlobby.commands

import com.mojang.brigadier.Command
import com.mojang.brigadier.builder.LiteralArgumentBuilder
import com.velocitypowered.api.command.BrigadierCommand
import com.velocitypowered.api.command.CommandSource
import com.velocitypowered.api.permission.Tristate
import com.velocitypowered.api.proxy.Player
import me.adrian3d.vlobby.VLobby
import me.adrian3d.vlobby.enums.SendMode
import me.adrian3d.vlobby.extensions.sendMiniMessage
import net.kyori.adventure.text.minimessage.tag.resolver.Placeholder

fun loadCommand(plugin: VLobby) {
    val aliases = plugin.config.commands
    if (aliases.isEmpty()) {
        return
    }
    val lobbyCommand = LiteralArgumentBuilder.literal<CommandSource>(aliases[0])
        .requires { it.getPermissionValue("vlobby.command") != Tristate.FALSE }
        .executes { cmd ->
            val source = cmd.source
            if (source !is Player) {
                plugin.logger.info(plugin.config.messages.consoleMessage)
                return@executes Command.SINGLE_SUCCESS
            }
            val mode: SendMode = plugin.config.servers.sendMode
            var lobbyToSend = mode.getServer(plugin)
            if (lobbyToSend == null)
                lobbyToSend =
                    if (mode == SendMode.RANDOM)
                        SendMode.FIRST_AVAILABLE.getServer(plugin)
                    else
                        SendMode.RANDOM.getServer(plugin)

            if (lobbyToSend == null) {
                plugin.logger.error("Cannot found lobby server to send")
                source.sendMiniMessage(plugin.config.messages.notAvailableServerMessage)
                return@executes Command.SINGLE_SUCCESS
            }
            source.createConnectionRequest(lobbyToSend).connect()
                .thenAccept { result ->
                    val server = result.attemptedConnection
                    if (result.isSuccessful) {
                        source.sendMiniMessage(
                            plugin.config.messages.successfullyMessage,
                            Placeholder.unparsed(
                                "server",
                                server.serverInfo.name
                            )
                        )
                    } else {
                        plugin.logger.error("Cannot connect to server: {}", server.serverInfo.name)
                        source.sendMiniMessage(
                            plugin.config.messages.errorMessage,
                            Placeholder.unparsed(
                                "server",
                                server.serverInfo.name
                            )
                        )
                    }
                }
            Command.SINGLE_SUCCESS
        }
        .build()
    val bCommand = BrigadierCommand(lobbyCommand)
    val manager = plugin.proxy.commandManager
    val builder = manager.metaBuilder(bCommand).plugin(plugin)
    if (aliases.size > 1) {
        builder.aliases(*aliases.toTypedArray().copyOfRange(1, aliases.size))
    }
    manager.register(builder.build(), bCommand)
}
